### PR TITLE
[handle-dyno-death] Try to get job-cancel event from API to work in client

### PR DESCRIPTION
**Pull request by wlonk**
See https://trello.com/c/PJcbJsi7/35-05-as-an-admin-i-dont-want-to-be-left-hanging-if-heroku-screws-me-up

Spike

### CODE REVIEW

File: src/js/socket/reducer.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
+// @flow
+
+import type { SocketAction } from 'socket/actions';
+
+export type Socket = {
+  +connected: boolean,
+} | null;
```
Comment (jgerigmeyer) [inline src/js/socket/reducer.js]: Is there a reason for the `null` state? Maybe simpler to always have it there, and just default to `{ connected: false }`.

File: src/js/app/reducer.js [hunk 2]
```diff
@@ -7,20 +7,23 @@ import org from 'org/reducer';
 import preflights from 'plans/reducer';
 import products from 'products/reducer';
 import user from 'user/reducer';
+import socket from 'socket/reducer';
```
Comment (jgerigmeyer) [inline src/js/app/reducer.js]: Alphabetize imports (throughout).

File: src/js/components/header/index.js [hunk 3]
```diff
@@ -50,12 +53,20 @@ const Header = ({ user, doLogout }: Props) => (
         {user ? <Logout user={user} doLogout={doLogout} /> : <Login />}
       </>
     }
+    info={
+      socket && socket.connected ? null : (
+        <>
+          <span>You&#39;re in offline mode.</span>
+        </>
+      )
+    }
```
Comment (jgerigmeyer) [inline src/js/components/header/index.js]: I assume this is just a placeholder? Based on the mockup I think we want an Alert component: https://react.lightningdesignsystem.com/components/alerts/

File: src/js/socket/reducer.js (hunk 1 above)
Comment (jgerigmeyer) [reply src/js/socket/reducer.js]:
```suggestion
};
```

File: src/js/socket/reducer.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+const reducer = (socket: Socket = null, action: SocketAction): Socket => {
```
Comment (jgerigmeyer) [inline src/js/socket/reducer.js]:
```suggestion
const reducer = (socket: Socket = { connected: false }, action: SocketAction): Socket => {
```

File: src/js/components/header/index.js (hunk 3 above)
Comment (wlonk) [reply src/js/components/header/index.js]: Yep, it's a placeholder; the whole commit kinda is, more of a "can I get this to work" than not. You'll note it didn't have tests attached, either, for the same reason.

File: metadeploy/api/jobs.py [hunk 5]
```diff
@@ -265,7 +265,7 @@ def expire_user_tokens():
 
 def preflight(preflight_result_id):
     # Because the FieldTracker interferes with transparently serializing models across
-    # the Redis boundary, we have to pass a primative value to this function,
+    # the Redis boundary, we have to pass a primitive value to this function,
```
Comment (wlonk) [inline metadeploy/api/jobs.py]: I am embarrassed by this typo!

File: src/js/components/plans/jobResults.js [hunk 6]
```diff
@@ -105,10 +105,13 @@ const JobResults = ({
     currentJob.error_count !== undefined && currentJob.error_count > 0;
   const hasWarnings =
     currentJob.warning_count !== undefined && currentJob.warning_count > 0;
+  const canceledPreflight =
+    preflight && preflight && currentJob.status === CONSTANTS.STATUS.CANCELED;
```
Comment (wlonk) [inline src/js/components/plans/jobResults.js]: `preflight && preflight` seems redundant.

File: src/js/utils/websockets.js [hunk 7]
```diff
@@ -156,11 +166,18 @@ export const createSocket = ({
...
     onclose: e => {
       log('[WebSocket] closed');
-      open = false;
+      if (open) {
+        open = false;
+        setTimeout(() => {
+          if (!open) {
+            dispatch(disconnectSocket());
+          }
+        }, 5000);
```
[the remark this answers is not in the document]
Comment (wlonk) [reply src/js/utils/websockets.js]: That very much works for me.

File: src/js/components/plans/jobResults.js (hunk 6 above)
Comment (jgerigmeyer) [reply src/js/components/plans/jobResults.js]: Never can be too sure!

**Comment by wlonk** [comment]
I'll merge when it's 💚 .

**Pull request merged**